### PR TITLE
Allow flexible mask shapes in splitting loss + fix splitting losses for multicoil MRI

### PR DESCRIPTION
**Pull request by Andrewwango**
- Allow flexible mask shapes in splitting loss (e.g. for FastMRI data)
- Fix splitting losses for multicoil MRI (i.e. where there is an extra dim in y but not in mask)
- Fix backward pass for least squares when gamma is batched but y.ndim !=x.ndim

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/loss/mri/measplit.py [hunk 1]
```diff
@@ -70,24 +71,48 @@ class WeightedSplittingLoss(SplittingLoss):
...
 
         def forward(self, y1, y2):
             """Weighted metric forward pass."""
+
+            if y1.shape[-2:] != y2.shape[-2:]:
+                raise ValueError(
+                    f"Metric input tensors should be same image size but got {y1.shape[-2:]}, {y2.shape[-2:]}."
+                )
+
+            if self.weight.shape[-1] != y1.shape[-1]:
```
Comment (Andrewwango) [inline deepinv/loss/mri/measplit.py]: this is going to lead to some overhead at train time when you iterate over a dataset with different image sizes. A hacky solution would be to cache weights per width. I think it's ok for now to leave it. At least it's ok for single-image training.

Comment (matthieutrs) [reply deepinv/loss/mri/measplit.py]: Agreed.
Another option is to add a shape parameter at init of the loss to force the user acknowledging that shape should always be the same.
(still not super elegant, I don't necessarily think it's better, current solution is good IMO)

File: deepinv/loss/mri/measplit.py [hunk 2]
```diff
@@ -193,14 +237,34 @@ def __init__(
         self.noise_model = noise_model
         self.noise_model.update_parameters(sigma=noise_model.sigma * alpha)
 
+    @staticmethod
+    def expand_mask(mask, y):
+        """Expand mask intermediate dimensions to match those of y, where intermediate
+        dimensions are those (e.g. slice, coils, time) that are not the first two dims (batch, channel),
```
Comment (matthieutrs) [inline deepinv/loss/mri/measplit.py]:
```suggestion
        dimensions are those (e.g. depth, coils, time) that are not the first two dims (batch, channel),
```

File: deepinv/loss/mri/measplit.py [hunk 3]
```diff
@@ -193,14 +237,34 @@ def __init__(
...
         recon_loss = super().forward(x_net, y, physics, model, **kwargs)
 
         mask = model.get_mask() * getattr(physics, "mask", 1.0)  # M_\lambda\cap\omega
+
         n2n_metric = self.WeightedMetric(
-            (1 + 1 / (self.alpha**2)) * mask, self.metric.metric, expand=False
+            mask_generator=None,
+            physics_generator=None,
+            weight=(1 + 1 / (self.alpha**2)) * self.expand_mask(mask, y),
+            metric=self.metric.metric,
```
Comment (matthieutrs) [inline deepinv/loss/mri/measplit.py]:
```suggestion
            metric=self.metric.metric,
```
unelated to your proposed changes but I am wondering if we should not think of a better naming to avoid `self.metric.metric`

File: deepinv/physics/generator/inpainting.py [hunk 4]
```diff
@@ -91,13 +96,17 @@ def step(
 
         :param int batch_size: batch_size. If None, no batch dimension is created. If input_mask passed and has its own batch dimension > 1, batch_size is ignored.
         :param torch.Tensor, None input_mask: optional mask to be split. If None, all pixels are considered. If not None, only pixels where mask==1 are considered. input_mask shape can optionally include a batch dimension.
+        :param tuple img_size: optionally reset the 2D image size on-the-fly, must be of form (H, W).
```
Comment (matthieutrs) [inline deepinv/physics/generator/inpainting.py]: I think this sentence could be clarified: what exactly is meant by "reset the 2D image size on the fly"? Does it mean "reset the attribute self.img_size to the img_size value on the fly"? The way it is formulated now here suggests to me that `img_size` should be a bool.
Maybe it would be worth adding an example with this param used around line 40 above?

File: deepinv/physics/generator/inpainting.py [hunk 5]
```diff
@@ -163,13 +174,20 @@ def check_pixelwise(self, input_mask=None) -> bool:
...
 
-    def batch_step(self, input_mask: torch.Tensor = None) -> dict:
+    def batch_step(
+        self, input_mask: torch.Tensor = None, img_size: Optional[tuple] = None
+    ) -> dict:
         r"""
         Create one batch of splitting mask.
 
         :param torch.Tensor, None input_mask: optional mask to be split. If ``None``, all pixels are considered. If not ``None``, only pixels where ``mask==1`` are considered. Batch dimension should not be included in shape.
+        :param tuple img_size: optionally reset the 2D image size on-the-fly, must be of form (H, W).
```
Comment (matthieutrs) [inline deepinv/physics/generator/inpainting.py]: same as above

File: deepinv/physics/generator/inpainting.py [hunk 6]
```diff
@@ -262,16 +280,26 @@ def __init__(
         )
         self.split_generator = split_generator
 
-    def batch_step(self, input_mask: torch.Tensor = None) -> dict:
-        mask = self.split_generator.step(batch_size=1)["mask"].squeeze(0)
+    def batch_step(
+        self, input_mask: torch.Tensor = None, img_size: Optional[tuple] = None
+    ) -> dict:
+
```
Comment (matthieutrs) [inline deepinv/physics/generator/inpainting.py:295]: add docstrings?

File: deepinv/physics/generator/inpainting.py [hunk 7]
```diff
@@ -474,15 +508,22 @@ def __init__(
...
+        self, input_mask: torch.Tensor = None, img_size: Optional[tuple] = None
+    ) -> dict:
         if len(self.img_size) != 4:
-            raise ValueError("img_size must be of shape (C, T, H, W)")
+            raise ValueError("Default img_size must be of shape (C, T, H, W)")
 
-        if not isinstance(input_mask, torch.Tensor) or input_mask.numel() <= 1:
-            input_mask = torch.ones(self.img_size, device=self.device)
+        if input_mask is not None and input_mask.shape != self.img_size:
+            raise ValueError("input_mask must be same shape as default img_size")
```
Comment (matthieutrs) [inline deepinv/physics/generator/inpainting.py:543]:
```suggestion
        if input_mask is not None:
           if input_mask.shape != self.img_size and img_size is None or img_size!=input_mask.shape:
            raise ValueError(f"input_mask must be same shape as default img_size. Got input_mask.shape {input_mask.shape:2e} and img_size {img_size:2e}.")
```
shouldnt there be a potential check on whether to use img_size or self.img_size?

File: deepinv/physics/generator/inpainting.py [hunk 8]
```diff
@@ -522,7 +563,10 @@ def __init__(
         self.prev_split_size = None
 
     def batch_step(
-        self, input_mask: torch.Tensor = None, persist_prev: bool = False
+        self,
+        input_mask: torch.Tensor = None,
+        img_size: Optional[tuple] = None,
+        persist_prev: bool = False,
```
Comment (matthieutrs) [inline deepinv/physics/generator/inpainting.py:592]: add docstrings

File: deepinv/tests/test_loss.py [hunk 9]
```diff
@@ -460,6 +492,11 @@ def test_measplit(device, loss_name, rng):
             eval_n_samples=eval_n_samples,
         )
     elif loss_name == "splitting_eval_split_input_output":
+        if physics_name == "MultiCoilMRI":
+            pytest.skip(
```
Comment (matthieutrs) [inline deepinv/tests/test_loss.py]: Is it raising an error in this case? If yes, maybe it'd be good to check that it actually raises an error rather than just skipping (to avoid having something that works but shouldnt)

**Review by matthieutrs — CHANGES_REQUESTED** [review]
Thanks a lot for this PR!! Just a few minor comments, mainly about docs

File: deepinv/loss/mri/measplit.py (hunk 1 above)
Comment (Andrewwango) [reply deepinv/loss/mri/measplit.py]: added some comments to make it clearer what is going on

File: deepinv/loss/mri/measplit.py (hunk 3 above)
Comment (Andrewwango) [reply deepinv/loss/mri/measplit.py]: Yeah to be honest it's not optimal but I can't think of a better solution

File: deepinv/physics/generator/inpainting.py (hunk 4 above)
Comment (Andrewwango) [reply deepinv/physics/generator/inpainting.py]: good point, thanks

File: deepinv/tests/test_loss.py (hunk 9 above)
Comment (Andrewwango) [reply deepinv/tests/test_loss.py]: upon second thought, the test should be fine - so I removed the skip!

File: deepinv/physics/generator/inpainting.py (hunk 8 above)
Comment (Andrewwango) [reply deepinv/physics/generator/inpainting.py:592]: thanks, I also added a img_size check

File: deepinv/physics/generator/inpainting.py (hunk 6 above)
Comment (Andrewwango) [reply deepinv/physics/generator/inpainting.py:295]: done

File: deepinv/physics/generator/inpainting.py (hunk 7 above)
Comment (Andrewwango) [reply deepinv/physics/generator/inpainting.py:543]: No, because only one of `img_size`  or `input_mask` should be specified, or None. If both are specified, then an error is raised in `step` (which calls `batch_step`)

File: deepinv/loss/mri/measplit.py [hunk 10]
```diff
@@ -69,24 +72,49 @@ class WeightedSplittingLoss(SplittingLoss):
...
         def forward(self, y1, y2):
             """Weighted metric forward pass."""
+
+            if y1.shape[-2:] != y2.shape[-2:]:
+                raise ValueError(
+                    f"Metric input tensors should be same image size but got {y1.shape[-2:]}, {y2.shape[-2:]}."
+                )
+
+            if self.weight.shape[-1] != y1.shape[-1]:
+                # TODO these weights should be cached for a speed-up
```
Comment (tachella) [inline deepinv/loss/mri/measplit.py]: this has been done?

File: deepinv/loss/mri/measplit.py (hunk 3 above)
Comment (tachella) [reply deepinv/loss/mri/measplit.py]: I would rename the `metric` method of  `WeightedMetric` as `forward` to avoid this

**Review by tachella — APPROVED** [review]
LGTM, minor modifs

File: deepinv/loss/mri/measplit.py (hunk 3 above)
Comment (Andrewwango) [reply deepinv/loss/mri/measplit.py]: renamed to pixel_metric

File: deepinv/loss/mri/measplit.py (hunk 10 above)
Comment (Andrewwango) [reply deepinv/loss/mri/measplit.py]: done

File: deepinv/optim/utils.py [hunk 11]
```diff
@@ -900,11 +900,16 @@ def forward(
 
         # Save tensors only
         gamma_orig_shape = gamma.shape
-        # For broadcasting with other tensors
-        if gamma.ndim > 0:
-            gamma = gamma.view(
-                -1, *[1] * (y.ndim - 1)
-            )  # reshape for broadcasting if needed
+        # For broadcasting with other tensors. Note we already have checked gamma shapes
```
Comment (tachella) [inline deepinv/optim/utils.py:903]: looks good to me
